sound/soc/sof/ipc4-topology.c

```diff
@@ -939,8 +984,67 @@ static void sof_ipc4_widget_free_comp_process(struct snd_sof_widget *swidget)
 	swidget->private = NULL;
 }
 
+static void
+sof_ipc4_update_cpc_from_tplg(struct snd_sof_dev *sdev,
+			      struct snd_sof_widget *swidget,
+			      struct sof_ipc4_module_params *module_params,
+			      struct sof_ipc4_base_module_cfg *basecfg,
+			      bool exact_match)
+{
+	struct sof_ipc4_tplg_mod_cfg *tplg_mod_cfgs = module_params->mod_cfgs;
+	u32 cpc_pick = 0;
+	u32 max_cpc = 0;
+	int i;
+
+	/*
+	 * No mod_cfg present for the module or the CPC is already set, no need
+	 * to proceed further
+	 */
+	if (!tplg_mod_cfgs || basecfg->cpc)
+		return;
+
+	/*
+	 * Find the best matching (highest) CPC value based on the module's
+	 * external configuration (IBS/OBS).
+	 * The CPC value in each module config entry has been measured and
+	 * recorded as a IBS/OBS/CPC triplet.
```

Comment:
what if there is not ibs?

or what if there is no obs?

The assumption that all modules have input and output buffers does not make any sense to me.

Reply:
If no ibs, then ibs is 0, right?
I don't see any problem with that.
The array in manifest is created based on the measured configuration, so ibs=0, the module have ibs=0, so that part is constant.